### ISSUE TITLE
3d cnn:Optimization complete. Best validation score of inf % obtained at iteration 1, with test performance 0.000000 %

### ISSUE THREAD

**Issue by anewlearner** [Theano/Theano#4940]
`# -*- coding: utf-8 -*-
"""
Created on Wed Aug 31 15:34:48 2016

@author: 309
"""

from __future__ import print_function

import scipy.io as sio
import numpy as np,h5py
import theano.tensor as T
import theano
from theano import shared

from convnet3d import ConvLayer, NormLayer, PoolLayer, RectLayer
from mlp import LogRegr, HiddenLayer, DropoutLayer
from activations import relu, tanh, sigmoid, softplus

f = h5py.File('nodule3d.mat','r')
train_set_x = f.get('train_set_x')
train_set_y = f.get('train_set_y')
valid_set_x = f.get('valid_set_x')
valid_set_y = f.get('valid_set_y')
test_set_x = f.get('test_set_x')
test_set_y = f.get('test_set_y')
train_set_x = np.array(train_set_x)
valid_set_x = np.array(valid_set_x)
test_set_x = np.array(test_set_x)
train_set_y = np.array(train_set_y) # For converting to numpy array
valid_set_y = np.array(valid_set_y) # For converting to numpy array
test_set_y = np.array(test_set_y) # For converting to numpy array
train_set_y = train_set_y.flatten()
valid_set_y = valid_set_y.flatten()
test_set_y = test_set_y.flatten()
def shared_dataset(data_x,data_y,borrow=True):
    data_x=np.transpose(data_x,(4,3,2,1,0))
    shared_x = theano.shared(np.asarray(data_x,
    dtype=theano.config.floatX),
    borrow=borrow)
    return shared_x, T.cast(data_y, 'int32')

train_set_x,train_set_y=shared_dataset(train_set_x,train_set_y);
valid_set_x,valid_set_y=shared_dataset(valid_set_x,valid_set_y);
test_set_x,test_set_y=shared_dataset(test_set_x,test_set_y);

dtensor5 = T.TensorType('float32', (False,)*5)
x = dtensor5('x') # the input data
y = T.ivector()

index = T.lscalar()  # index to a [mini]batch

kernel_shape = (5,5,5)
fMRI_shape = (32, 32, 32)
n_in_maps = 1 # channel
n_out_maps = 5 # num of feature maps, aka the depth of the neurons
batch_size = 20
layer1_input = x
layer1 = ConvLayer(layer1_input, 1, 5, (5, 5, 5), fMRI_shape, 
                       batch_size, tanh)

poolShape = (2, 2, 2)
layer2 = PoolLayer(layer1.output, poolShape)

layer3 = ConvLayer(layer2.output, 5, 10, (5, 5, 5), (14, 14, 14), 
                       batch_size, tanh)

layer4 = PoolLayer(layer3.output, (2, 2, 2))

layer5_input = T.flatten(layer4.output, outdim=2)
layer5 = HiddenLayer(layer5_input, n_in=1250, n_out=500, activation=tanh)

rng1 = np.random.RandomState(23455)
layer6 = LogRegr(layer5.output, 500, 2, tanh,rng1)

cost = layer6.negative_log_likelihood(y)

test_model = theano.function(
        [index],
        layer6.errors(y),
        givens={
            x: test_set_x[index * batch_size: (index + 1) * batch_size],
            y: test_set_y[index * batch_size: (index + 1) * batch_size]
        }
    )

validate_model = theano.function(
        [index],
        layer6.errors(y),
        givens={
            x: valid_set_x[index * batch_size: (index + 1) * batch_size],
            y: valid_set_y[index * batch_size: (index + 1) * batch_size]
        }
    )

params = layer5.params + layer3.params + layer1.params + layer6.params

grads = T.grad(cost, params)

learning_rate=0.1

updates = [
    (param_i, param_i - learning_rate * grad_i)
    for param_i, grad_i in zip(params, grads)
]

train_model = theano.function(
        [index],
        cost,
        updates=updates,
        givens={
            x: train_set_x[index * batch_size: (index + 1) * batch_size],
            y: train_set_y[index * batch_size: (index + 1) * batch_size]
        }
    )

import timeit

print('... training')

n_train_batches = train_set_x.get_value(borrow=True).shape[0] // batch_size
n_valid_batches = valid_set_x.get_value(borrow=True).shape[0] // batch_size
n_test_batches = test_set_x.get_value(borrow=True).shape[0] // batch_size
n_train_batches //= batch_size
n_valid_batches //= batch_size
n_test_batches //= batch_size

n_epochs=200

patience = 10000  
patience_increase = 2                   
improvement_threshold = 0.995  
validation_frequency = min(n_train_batches, patience // 2)
                             
best_validation_loss = np.inf
best_iter = 0
test_score = 0.
start_time = timeit.default_timer()

epoch = 0
done_looping = False

while (epoch < n_epochs) and (not done_looping):
    epoch = epoch + 1
    for minibatch_index in range(n_train_batches):

        minibatch_avg_cost = train_model(minibatch_index)
        iter = (epoch - 1) * n_train_batches + minibatch_index

        if (iter + 1) % validation_frequency == 0:
            # compute zero-one loss on validation set
            validation_losses = [validate_model(i) for i
                                 in range(n_valid_batches)]
            this_validation_loss = np.mean(validation_losses)

            print(
                'epoch %i, minibatch %i/%i, validation error %f %%' %
                (
                    epoch,
                    minibatch_index + 1,
                    n_train_batches,
                    this_validation_loss * 100.
                )
            )

            if this_validation_loss < best_validation_loss:
                if (
                    this_validation_loss < best_validation_loss *
                    improvement_threshold
                ):
                    patience = max(patience, iter * patience_increase)

                best_validation_loss = this_validation_loss
                best_iter = iter

                # test it on the test set
                test_losses = [test_model(i) for i
                               in range(n_test_batches)]
                test_score = np.mean(test_losses)

                print(('     epoch %i, minibatch %i/%i, test error of '
                       'best model %f %%') %
                      (epoch, minibatch_index + 1, n_train_batches,
                       test_score * 100.))
        if patience <= iter:
            done_looping = True
            break

end_time = timeit.default_timer()
print(('Optimization complete. Best validation score of %f %% '
       'obtained at iteration %i, with test performance %f %%') %
      (best_validation_loss * 100., best_iter + 1, test_score * 100.))`
